packages/data-context/src/actions/NotificationActions.ts

```diff
@@ -0,0 +1,95 @@
+import type { NotifyWhenRunCompletes } from '@packages/types/src/preferences'
+import type { DataContext } from '..'
+import debugLib from 'debug'
+import assert from 'assert'
+import type { RelevantRunInfo } from '../gen/graphcache-config.gen'
+
+const debug = debugLib('cypress:data-context:NotificationActions')
+
+export class NotificationActions {
+  constructor (private ctx: DataContext) { }
+
+  async onNotificationClick (runNumber: number) {
+    debug('notification clicked for #%s', runNumber)
+    await this.ctx.actions.browser.focusActiveBrowserWindow()
+
+    await this.ctx.actions.project.debugCloudRun(runNumber)
+  }
+
+  private get projectTitle () {
+    assert(this.ctx.coreData.currentProject, 'current project needs to be set to send notifications')
+
+    return this.ctx.project.projectTitle(this.ctx.coreData.currentProject)
+  }
+
+  private get desktopNotificationsEnabledPreference () {
+    return this.ctx.coreData.localSettings.preferences.desktopNotificationsEnabled
+  }
+
+  private get notifyWhenRunStartsPreference () {
+    return this.ctx.coreData.localSettings.preferences.notifyWhenRunStarts
+  }
+
+  private get notifyWhenRunStartsFailingPreference () {
+    return this.ctx.coreData.localSettings.preferences.notifyWhenRunStartsFailing
+  }
+
+  private get notifyWhenRunCompletesPreference () {
+    return this.ctx.coreData.localSettings.preferences.notifyWhenRunCompletes
+  }
+
+  maybeSendRunNotification (cachedRun: RelevantRunInfo, newRun: RelevantRunInfo) {
+    if (this.desktopNotificationsEnabledPreference !== true) {
+      debug('desktopNotificationsEnabled not true, skipping notification for run #%s', newRun.runNumber)
+
+      return
+    }
+
+    switch (newRun.status) {
+      case 'RUNNING':
+        // If the new run has the same run number and last time we saw the run it had 0 failures and now it has more than 0
+        // failures, then it just started failing.
+        if ((cachedRun.runNumber === newRun.runNumber) && (cachedRun.totalFailed === 0 && newRun.totalFailed > 0)) {
+          this.sendRunFailingNotification(newRun.runNumber)
+        } else {
+          this.sendRunStartedNotification(newRun.runNumber)
+        }
+
+        break
+      default:
+        if (newRun.status) {
```

Comment:
why do we need this `if` here? Won't it never get here in the first place without a `newRun.status` since thats the switch case?

Reply:
I'm guessing `switch('PASSED')`, or anything other than `RUNNING` would mean we fall through to this case.

It looks like we could simply do

```js
if (!newRun.status) {
  return
}
```

before the switch statement.

Reply:
@jordanpowell88 if `newRun.status` is undefined we will still reach the default case. Yes, we could return early if there is no status

Reply:
https://github.com/cypress-io/cypress/pull/26974/commits/dcf4794f5ac4c2afda4f313fa270783b80035a82